elixir/lib/lexer.ex

```diff
@@ -68,84 +68,110 @@ defmodule Monkey.Lexer do
+    # it won't be as well optimised due to having less context on what the execution paths would be
+    #
+    # It is more efficient to prepend to a list and reverse it than to append to
+    # the list while building it. Lists here are linked-lists, so the whole list
+    # would be copied each time we append to it.
+    # See: https://www.erlang.org/doc/efficiency_guide/listhandling
+    #
+    # Uses binary pattern-matching to match on the first character(s).
+    # e.g. `<<c::8, rest::binary>>` matches on the first 8 bits and assigns it
+    # to `c`, then assigns the rest of the binary to `rest`.
+    # For more details, see: https://hexdocs.pm/elixir/Kernel.SpecialForms.html#<<>>/1
+    #
+    # this logic works equivalent if done as multiple function heads, or a case statement (the compiled BEAM code is
+    # identical either way). The benefit of a case statement is each case clause only has to worry about the string
+    # match, and doesn't have to cart around `tokens`, as we've already done that when we entered the function.
+    case input do
+      # Recursive base-case. When the input is empty, we add an EOF token.
+      <<>> -> [:eof | tokens] |> Enum.reverse()
+      # Ignore whitespace.
+      <<c::8, rest::binary>> when is_whitespace(c) -> lex(rest, tokens)
+      <<"==", rest::binary>> -> lex(rest, [:equal | tokens])
+      <<"!=", rest::binary>> -> lex(rest, [:not_equal | tokens])
+      <<";", rest::binary>> -> lex(rest, [:semicolon | tokens])
+      <<",", rest::binary>> -> lex(rest, [:comma | tokens])
+      <<"(", rest::binary>> -> lex(rest, [:lparen | tokens])
+      <<")", rest::binary>> -> lex(rest, [:rparen | tokens])
+      <<"{", rest::binary>> -> lex(rest, [:lsquirly | tokens])
+      <<"}", rest::binary>> -> lex(rest, [:rsquirly | tokens])
+      <<"=", rest::binary>> -> lex(rest, [:assign | tokens])
+      <<"+", rest::binary>> -> lex(rest, [:plus | tokens])
+      <<"-", rest::binary>> -> lex(rest, [:minus | tokens])
+      <<"!", rest::binary>> -> lex(rest, [:bang | tokens])
+      <<"/", rest::binary>> -> lex(rest, [:slash | tokens])
+      <<"*", rest::binary>> -> lex(rest, [:asterisk | tokens])
+      <<">", rest::binary>> -> lex(rest, [:greater_than | tokens])
+      <<"<", rest::binary>> -> lex(rest, [:less_than | tokens])
+      <<"fn", rest::binary>> -> maybe_keyword(rest, byte_size("fn"), input, :function, tokens)
+      <<"let", rest::binary>> -> maybe_keyword(rest, byte_size("let"), input, :let, tokens)
+      <<"if", rest::binary>> -> maybe_keyword(rest, byte_size("if"), input, :if, tokens)
+      <<"else", rest::binary>> -> maybe_keyword(rest, byte_size("else"), input, :else, tokens)
+      <<"true", rest::binary>> -> maybe_keyword(rest, byte_size("true"), input, true, tokens)
+      <<"false", rest::binary>> -> maybe_keyword(rest, byte_size("false"), input, false, tokens)
+      <<"return", rest::binary>> -> maybe_keyword(rest, byte_size("return"), input, :return, tokens)
```

Comment:
Does this effect benchmark at all?
```suggestion
      <<"fn", c::8, rest::binary>> when not is_letter(c) -> lex(<<c::8, rest::binary>>, [:function | tokens])
      <<"let", c::8, rest::binary>> when not is_letter(c) -> lex(<<c::8, rest::binary>>, [:let | tokens])
      <<"if", c::8, rest::binary>> when not is_letter(c) -> lex(<<c::8, rest::binary>>, [:if | tokens])
      <<"else", c::8, rest::binary>> when not is_letter(c) -> lex(<<c::8, rest::binary>>, [:else | tokens])
      <<"true", c::8, rest::binary>> when not is_letter(c) -> lex(<<c::8, rest::binary>>, [true | tokens])
      <<"false", c::8, rest::binary>> when not is_letter(c) -> lex(<<c::8, rest::binary>>, [false | tokens])
      <<"return", c::8, rest::binary>> when not is_letter(c) -> lex(<<c::8, rest::binary>>, [:return | tokens])
```